src/toast/_libtoast/accelerator.cpp

```diff
@@ -0,0 +1,401 @@
+            # else // ifdef USE_OPENACC_MEMPOOL
+            acc_update_device(p_host, n_bytes);
+            # endif // ifdef USE_OPENACC_MEMPOOL
+            return;
+        }, py::arg(
+            "data"), R"(
+        Update device copy of the data from the host.
+
+        Args:
+            data (array):  The host data.
+
+        Returns:
+            None
+
+    )");
+
+    m.def(
+        "acc_update_self", [](py::buffer data) {
+            auto & log = toast::Logger::get();
+            py::buffer_info info = data.request();
+            void * p_host;
+            size_t n_elem;
+            size_t n_bytes;
+            extract_buffer_info(info, &p_host, &n_elem, &n_bytes);
+
+            bool present = acc_is_present(p_host, n_bytes);
+            if (!present) {
+                std::ostringstream o;
+                o << "Data is not present on device, cannot update host.";
+                log.error(o.str().c_str());
+                throw std::runtime_error(
+                          o.str().c_str());
+            }
+
+            std::ostringstream o;
+            o << "update host/self with host pointer " << p_host << " (" << n_bytes << " bytes)";
+            log.verbose(o.str().c_str());
+
+            # ifdef USE_OPENACC_MEMPOOL
+
+            // FIXME: Does the pool have an update device / host equivalent?
+            auto & pool = GPU_memory_pool::get();
+            pool.fromDevice(reinterpret_cast <char *> (p_host));
+            # else // ifdef USE_OPENACC_MEMPOOL
+            acc_update_self(p_host, n_bytes);
+            # endif // ifdef USE_OPENACC_MEMPOOL
+            return;
+        }, py::arg(
+            "data"), R"(
+        Update host copy of the data from the device.
+
+        Args:
+            data (array):  The host data.
+
+        Returns:
+            None
+
+    )");
+
+    m.def(
+        "acc_delete", [](py::buffer data) {
+            auto & log = toast::Logger::get();
+            py::buffer_info info = data.request();
+            void * p_host;
+            size_t n_elem;
+            size_t n_bytes;
+            extract_buffer_info(info, &p_host, &n_elem, &n_bytes);
+
+            bool present = acc_is_present(p_host, n_bytes);
+            if (!present) {
+                std::ostringstream o;
+                o << "Data is not present on device, cannot delete.";
+                log.error(o.str().c_str());
+                throw std::runtime_error(
+                          o.str().c_str());
+            }
+
+            std::ostringstream o;
+            o << "delete device mem for host pointer " << p_host << " (" << n_bytes << " bytes)";
+            log.verbose(o.str().c_str());
+
+            # ifdef USE_OPENACC_MEMPOOL
+
+            // FIXME:  If there was a pool.Free(cpu pointer), we would call that here.
+            auto & pool = GPU_memory_pool::get();
+            pool.fromDevice(reinterpret_cast <char *> (p_host));
```

Comment:
Semantically, `pool.fromDevice()` and `acc_delete()` do not seem equivalent. One implies retrieving the data from the card while other implies freeing the device memory. If `acc_delete()` also updates the host pointer, perhaps that should be part of the routine name: `acc_sync_and_delete()`.

Reply:
The pool currently does not provide some methods (update device and update host), and I agree that to match the syntax, the pool should eventually support a method that calls `Free` with a host pointer lookup.  Note that the names in this file are designed to match the function names in the OpenACC standard.

Reply:
What are the methods that would be needed? (I could add them this morning)
- free(cpu_ptr)
- update(cpu_ptr)

Reply:
Done!